libsel4vm/src/arch/arm/vgic/vgic_v2.c

```diff
@@ -0,0 +1,300 @@
+/*
+ * Copyright 2019, Data61, CSIRO (ABN 41 687 119 230)
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+/*
+ * This component controls and maintains the GIC for the VM.
+ * IRQs must be registered at init time with vm_virq_new(...)
+ * This function creates and registers an IRQ data structure which will be used for IRQ maintenance
+ * b) ENABLING: When the VM enables the IRQ, it checks the pending flag for the VM.
+ *   - If the IRQ is not pending, we either
+ *        1) have not received an IRQ so it is still enabled in seL4
+ *        2) have received an IRQ, but ignored it because the VM had disabled it.
+ *     In either case, we simply ACK the IRQ with seL4. In case 1), the IRQ will come straight through,
+       in case 2), we have ACKed an IRQ that was not yet pending anyway.
+ *   - If the IRQ is already pending, we can assume that the VM has yet to ACK the IRQ and take no further
+ *     action.
+ *   Transitions: b->c
+ * c) PIRQ: When an IRQ is received from seL4, seL4 disables the IRQ and sends an async message. When the VMM
+ *    receives the message.
+ *   - If the IRQ is enabled, we set the pending flag in the VM and inject the appropriate IRQ
+ *     leading to state d)
+ *   - If the IRQ is disabled, the VMM takes no further action, leading to state b)
+ *   Transitions: (enabled)? c->d :  c->b
+ * d) When the VM acknowledges the IRQ, an exception is raised and delivered to the VMM. When the VMM
+ *    receives the exception, it clears the pending flag and acks the IRQ with seL4, leading back to state c)
+ *    Transition: d->c
+ * g) When/if the VM disables the IRQ, we may still have an IRQ resident in the GIC. We allow
+ *    this IRQ to be delivered to the VM, but subsequent IRQs will not be delivered as seen by state c)
+ *    Transitions g->c
+ *
+ *   NOTE: There is a big assumption that the VM will not manually manipulate our pending flags and
+ *         destroy our state. The affects of this will be an IRQ that is never acknowledged and hence,
+ *         will never occur again.
+ */
+
+#include "vgic.h"
+
+#include <stdint.h>
+#include <string.h>
+#include <stdlib.h>
+
+#include <utils/arith.h>
+#include <vka/vka.h>
+#include <vka/capops.h>
+
+#include <sel4vm/gen_config.h>
+#include <sel4vm/guest_vm.h>
+#include <sel4vm/boot.h>
+#include <sel4vm/guest_memory.h>
+#include <sel4vm/guest_irq_controller.h>
+#include <sel4vm/guest_vm_util.h>
+
+#include "vgicv2_defs.h"
+#include "vm.h"
+#include "../fault.h"
+#include "virq.h"
+#include "gicv2.h"
+#include "vdist.h"
+
+
+static struct vgic_dist_device *vgic_dist;
+
+static inline struct gic_dist_map *vgic_priv_get_dist(struct vgic_dist_device *d)
+{
+    assert(d);
+    assert(d->priv);
+
+    vgic_t *vgic = vgic_device_get_vgic(d);
+    return priv_get_dist(vgic->dist);
```

Comment:
```suggestion
    return vgic->dist;
```
This seems a merge artifact.

Reply:
Thanks for spotting. I moved the priv_get_dist changes into a later commit so that the move commits didn't also modify the code.